compose_api.go

```diff
@@ -0,0 +1,290 @@
+	// RemoveImagesLocal - remove only images that don't have a tag
+	RemoveImagesLocal
+)
+
+type dockerComposeAPI struct {
+	lock           sync.RWMutex
+	name           string
+	configs        []string
+	waitStrategies map[string]wait.Strategy
+	containers     map[string]*DockerContainer
+	composeService api.Service
+	dockerClient   client.APIClient
+	projectOptions []cli.ProjectOptionsFn
+	project        *types.Project
+}
+
+func (d *dockerComposeAPI) Services() []string {
+	return d.project.ServiceNames()
+}
+
+func (d *dockerComposeAPI) Down(ctx context.Context, opts ...StackDownOption) error {
+	options := stackDownOptions{
+		DownOptions: api.DownOptions{
+			Project: d.project,
+		},
+	}
+
+	for i := range opts {
+		opts[i].applyToStackDown(&options)
+	}
+
+	return d.composeService.Down(ctx, d.name, options.DownOptions)
+}
+
+func (d *dockerComposeAPI) Up(ctx context.Context, opts ...StackUpOption) (err error) {
+	d.project, err = d.compileProject()
+	if err != nil {
+		return err
+	}
+
+	upOptions := stackUpOptions{
+		CreateOptions: api.CreateOptions{
+			Services:             d.project.ServiceNames(),
+			Recreate:             api.RecreateDiverged,
+			RecreateDependencies: api.RecreateDiverged,
+		},
+		StartOptions: api.StartOptions{
+			Project: d.project,
+		},
+	}
+
+	for i := range opts {
+		opts[i].applyToStackUp(&upOptions)
+	}
+
+	if len(upOptions.Services) != len(d.project.Services) {
+		sort.Strings(upOptions.Services)
+
+		filteredServices := make(types.Services, 0, len(d.project.Services))
+
+		for i := range d.project.Services {
+			if idx := sort.SearchStrings(upOptions.Services, d.project.Services[i].Name); idx < len(upOptions.Services) && upOptions.Services[idx] == d.project.Services[i].Name {
+				filteredServices = append(filteredServices, d.project.Services[i])
+			}
+		}
+
+		d.project.Services = filteredServices
+	}
+
+	err = d.composeService.Up(ctx, d.project, api.UpOptions{
+		Create: upOptions.CreateOptions,
+		Start:  upOptions.StartOptions,
+	})
+
+	if len(d.waitStrategies) == 0 || err != nil {
+		return err
+	}
+
+	errGrp, errGrpCtx := errgroup.WithContext(ctx)
+
+	for svc, strategy := range d.waitStrategies { // pinning the variables
+		svc := svc
+		strategy := strategy
+
+		errGrp.Go(func() error {
+			target, err := d.ServiceContainer(errGrpCtx, svc)
+			if err != nil {
+				return err
+			}
+			return strategy.WaitUntilReady(errGrpCtx, target)
+		})
+	}
+
+	return errGrp.Wait()
+}
+
+func (d *dockerComposeAPI) WaitForService(s string, strategy wait.Strategy) ComposeStack {
+	d.waitStrategies[s] = strategy
+	return d
+}
+
+func (d *dockerComposeAPI) WithEnv(m map[string]string) ComposeStack {
+	d.projectOptions = append(d.projectOptions, withEnv(m))
+	return d
+}
+
+func (d *dockerComposeAPI) WithOsEnv() ComposeStack {
+	d.projectOptions = append(d.projectOptions, cli.WithOsEnv)
+	return d
+}
+
+func (d *dockerComposeAPI) ServiceContainer(ctx context.Context, svcName string) (*DockerContainer, error) {
```

Comment:
I think this should return the interface "Container" instead of DockerContainer

Reply:
Go's best practice is normally 'accept interfaces, return structs' therefore I think it should be fine to return the `*DockerContainer` and not `Container`